fdbserver/storageserver.actor.cpp

```diff
@@ -4906,6 +4967,421 @@ ACTOR Future<Void> validateRangeAgainstServers(StorageServer* data,
+		return std::make_pair(rangesA.front(), emptyRange);
+	}
+	TraceEvent(SevVerbose, "RangesSameBeforeSort").detail("RangesA", rangesA).detail("Rangesb", rangesB);
+	// sort in ascending order
+	std::sort(rangesA.begin(), rangesA.end(), [](KeyRange a, KeyRange b) { return a.begin < b.begin; });
+	std::sort(rangesB.begin(), rangesB.end(), [](KeyRange a, KeyRange b) { return a.begin < b.begin; });
+	TraceEvent(SevVerbose, "RangesSameAfterSort").detail("RangesA", rangesA).detail("Rangesb", rangesB);
+	ASSERT(elementsAreExclusiveWithEachOther(rangesA));
+	ASSERT(elementsAreExclusiveWithEachOther(rangesB));
+	if (rangesA.front().begin != rangesB.front().begin) { // rangeList heads mismatch
+		return std::make_pair(rangesA.front(), rangesB.front());
+	} else if (rangesA.back().end != rangesB.back().end) { // rangeList backs mismatch
+		return std::make_pair(rangesA.back(), rangesB.back());
+	}
+	int ia = 0;
+	int ib = 0;
+	KeyRangeRef rangeA = rangesA[0];
+	KeyRangeRef rangeB = rangesB[0];
+	KeyRange lastRangeA = Standalone(rangeA);
+	KeyRange lastRangeB = Standalone(rangeB);
+	while (true) {
+		if (rangeA.begin == rangeB.begin) {
+			if (rangeA.end == rangeB.end) {
+				if (rangeA.end == rangesA.back().end) {
+					break;
+				}
+				++ia;
+				++ib;
+				rangeA = rangesA[ia];
+				rangeB = rangesB[ib];
+				lastRangeA = Standalone(rangeA);
+				lastRangeB = Standalone(rangeB);
+			} else if (rangeA.end > rangeB.end) {
+				rangeA = KeyRangeRef(rangeB.end, rangeA.end);
+				++ib;
+				rangeB = rangesB[ib];
+				lastRangeB = Standalone(rangeB);
+			} else {
+				rangeB = KeyRangeRef(rangeA.end, rangeB.end);
+				++ia;
+				rangeA = rangesA[ia];
+				lastRangeA = Standalone(rangeA);
+			}
+		} else {
+			return std::make_pair(lastRangeA, lastRangeB);
+		}
+	}
+	return Optional<std::pair<KeyRange, KeyRange>>();
+}
+
+std::vector<KeyRange> getServerShardsByReadingShardInfo(StorageServer* data, Version readAtVersion) {
+	std::vector<KeyRange> res;
+	const UID serverID = data->thisServerID;
+	for (auto t : data->shards.intersectingRanges(allKeys)) {
+		TraceEvent(SevVerbose, "AuditStorageShardStorageServerShardCheckShardInfo", data->thisServerID)
+		    .detail("Range", t.value()->keys)
+		    .detail("AtVersion", readAtVersion)
+		    .detail("AuditServer", serverID)
+		    .detail("ReadWrite", t.value()->readWrite ? "True" : "False")
+		    .detail("Adding", t.value()->adding ? "True" : "False");
+		if (t.value()->assigned()) {
+			res.push_back(t.value()->keys);
+		}
+	}
+	return res;
+}
+
+ACTOR Future<std::vector<KeyRange>> getServerShardsByReadingServerKeys(StorageServer* data, Transaction* tr) {
+	state RangeResult readResult;
+	state std::vector<KeyRange> res;
+	state KeyRef keyBegin;
+	state int retryCount = 0;
+	state int readRangeCount = 0;
+	state KeyRangeRef range;
+	state const UID serverID = data->thisServerID;
+	loop {
+		try {
+			res.clear();
+			keyBegin = allKeys.begin;
+			while (true) {
```

Comment:
We need to read the serverKeys range in a single transaction, w/o resetting the transaction, my understanding is this `while(true)` loop is essentially not used. Can we try using a single round of krmGetRanges and ASSERT !res.more.

Reply:
Yes.